### ISSUE TITLE
input text sometimes doesn't send all the keys (revisited)

### ISSUE THREAD

**Issue by reyelts** [robotframework/SeleniumLibrary#1413]
It seems we are seeing the symptom reported by closed issue #1285. As requested in the original issue, I've attached a log file with --loglevel trace in effect. Here are the details...

When we try to use RobotFramework+SeleniumLibrary under Chrome, we lose characters that are sent to the web page as input text. Here's a simple example robot script that produces the symptom for us:

```
*** Settings ***
Documentation     Simple example using SeleniumLibrary.
Library           SeleniumLibrary

*** Variables ***
${BROWSER}        headlesschrome

*** Test Cases ***
My Test
    Open Browser        https://www.duckduckgo.com  ${BROWSER} 
    Set Selenium Speed  1s
    Input Text          search_form_input_homepage  This is a test
    Click Button        search_button_homepage
    Sleep               5s
    Title Should Be     garbage to force failure for screenshot
    [Teardown]          Close Browser

*** Keywords ***
```

This should cause the web page to search for "This is a test". However, only "Thi" makes it across to the web page (see attached screenshot).

We are running on Ubuntu 18.04 with the following:

```
$ apt list 2> /dev/null | grep google-chrome
google-chrome-beta/stable 76.0.3809.62-1 amd64
google-chrome-stable/stable,now 75.0.3770.100-1 amd64 [installed]
google-chrome-unstable/stable 77.0.3833.0-1 amd64
$ apt list 2> /dev/null | grep selenium
libtest-www-selenium-perl/bionic,bionic 1.36-2 all
python-selenium/bionic,bionic 3.8.0+dfsg1-3 all
python3-selenium/bionic,bionic 3.8.0+dfsg1-3 all
$ pip list 2> /dev/null | grep selenium
robotframework-selenium2library (3.0.0)
robotframework-seleniumlibrary (3.3.1)
selenium (3.141.0)
```

There are some rather baffling details:
- We see the identical symptom with both chrome and headlesschrome. 
- The same robot under firefox works perfectly.
- We do not see the symptom (on any browser) if we are logged onto the system locally or if we use ssh+X11 forwarding to another Linux system.
- We do see the symptom consistently on Mac machines using ssh+X11 forwarding to XQuartz.

If the problem were with XQuartz, we would expect the same symptom regardless of browser. If the problem were with the chrome/headlesschrome drivers, we would expect the same symptom when the X server is on Linux. The fact that others have seen the same symptom (issue #1285) has us looking here first.

Any ideas?

![chrome_issue_screenshot](https://user-images.githubusercontent.com/34944431/61004208-ab51ef80-a322-11e9-9322-7ecb0941f591.png)
[chrome_issue.log](https://github.com/robotframework/SeleniumLibrary/files/3379381/chrome_issue.log)

**Comment by aaltat** [comment]
Can not reproduce the problem, when the `Input Text` keyword is called, the logging always shows that keyword sends the full text to the browser driver. 
![input_text](https://user-images.githubusercontent.com/2665023/61982735-58588780-b006-11e9-9a6c-269c69eb0fac.png)

But I have an idea why the test sometimes fails. Most likely problem is related to the application state. I would assume that the application is not ready receive the interaction (in this case the inputting the text.).  This is most likely caused because JavaScript still is running in the background and modifying the DOM or each character triggers some sort of action and causes changes in the DOM. Therefore the full text is not seen in the application and the verification fails. 

This is something which can not be fixed in the SeleniumLibrary side but you should look at the libraries built on top of the SeleniumLibrary. Example if you application is build using AngularJS, looking then [looking to AngularJS](https://github.com/Selenium2Library/robotframework-angularjs) direction is useful. or perhaps [Selenium testability library](https://github.com/salabs/robotframework-seleniumtestability) solves your problem. 

But in any case, SeleniumLibrary is working as designed and because the solution to this particular can not build in to the SeleniumLibrary, I am closing this issue.

**Comment by reyelts** [comment]
Hmm... we can reproduce this issue at will. I modified the original sample to 1) wait for the target element to become enabled and then 2) wait another 5s before the Input Text is run. So, the field is for sure ready to receive input. However, it behaves the same as before... only part of the text makes it to the target element. I've attached the changed script below:

```
*** Settings ***
Documentation     Simple example using SeleniumLibrary.
Library           SeleniumLibrary

*** Variables ***
${BROWSER}        headlesschrome

*** Test Cases ***
My Test
    Open Browser        https://www.duckduckgo.com  ${BROWSER} 
    Wait Until Element Is Enabled  search_form_input_homepage
    Sleep               5s
    Input Text          search_form_input_homepage  This is a test
    Sleep               5s
    Capture Page Screenshot
    Click Button        search_button_homepage
    [Teardown]          Close Browser

*** Keywords ***
```

I suspect you are correct that there is some sort of timing issue. While robotframework-seleniumtestability seems like an option, it really isn't. The problem isn't that the field isn't ready for input. The second example I provided proves that... the field is more than ready. 

The problem case that triggered this for us is a simple userid/password field. We came up with the DuckDuckGo example so as to provide a generic case. Our real case is sending an email address and password. For that case, we're also waiting for the fields to be ready and then injected some seconds of extra time. Unlike the DDG example, we're missing characters in the middle of the input. For example, we're sending "myuserid@mycompany.com" and the web page is receiving something like "myu@mpany.com".

So, rather than the field not being ready... it seems, in some cases, Chrome cannot receive characters as fast as Selenium can send them. An online search reveals a number of hits. Most of them suggest sending one key at a time. Here's an example:

https://github.com/angular/protractor/issues/3196

In our experience and that of others, the symptom seems to be specific to Chrome for some reason. In our case, I'm guessing it has to do with how Chrome interacts with X servers. It appears that it is trying to push each character (or maybe a buffer of 2-3 characters) to the X server/display. And by the time it gets back for the next few characters incoming, they are somehow already lost.

The above example solution takes the different browser behaviors into account. It first tries to send everything at once and, if that fails, falls back to send one character at a time.

**Comment by aaltat** [comment]
I believe that you have a genuine problem. I also know that the problem is not in the SeleniumLibrary, because the keywords uses only [Selenium API](https://github.com/robotframework/SeleniumLibrary/blob/master/src/SeleniumLibrary/keywords/formelement.py#L424) and I believe that problem is somewhere in the downstream. By guessing blindly I would assume that when characters are received by your application, it triggers something in your application, which causes the application lose some characters in the input element. One way to test this out, is to create static HTML page, [like this one](https://github.com/robotframework/SeleniumLibrary/blob/master/atest/resources/html/forms/login.html) and see can you reproduce the problem.

**Comment by reyelts** [comment]
Test run as requested... and it demonstrates the symptom as already described. To perform the test, I ran Python's SimpleHTTPServer against the static HTML page you suggested and used this Robot Framework + SeleniumLibrary script below with --loglevel trace enabled. 

```
*** Settings ***
Documentation     Simple example using SeleniumLibrary.
Library           SeleniumLibrary

*** Variables ***
${BROWSER}        headlesschrome

*** Test Cases ***
My Test
    Open Browser        http://my.url.com:8000  ${BROWSER} 
    Wait Until Element Is Enabled  username_field
    Sleep               5s
    Input Text          username_field  This is a test
    Sleep               5s
    Capture Page Screenshot
   [Teardown]          Close Browser

*** Keywords ***
```

As you can see, the script just a revision of the original that pointed to DuckDuckGo, and it still sends the string "This is a test". On the browser side, all we get of that is "Thiiat" as shown on the attached screenshot. 

![image](https://user-images.githubusercontent.com/34944431/62077292-fc108600-b206-11e9-9e83-475f3e97d2de.png)

And the behavior is still limited to Chrome. Firefox gets the entire string. Also, both headlesschrome and regular chrome behave in the same manner. Here's a new log:

[chrome_issue_simple.log](https://github.com/robotframework/SeleniumLibrary/files/3444053/chrome_issue_simple.log)

**Comment by reyelts** [comment]
I spent more time playing around with this. Using the simple login page to obviate login issues, I tried the following:

- I changed formelement.py so _input_text_into_text_field() sends one character at a time (and added trace records to prove it). That had no effect... I get the same result as sending the whole string at once. 

- I tried changing "This is a test" to "tttt tt t tttt" (same string format... just all t's). All of the t's made it, but the spaces did not. It is almost as if something is seeing some of the characters as special.

- I switched the X server from XQuartz (Mac) with port forwarding over ssh to instead use xvfb locally on the Linux machine (where Robot Framework+SeleniumLibrary are running). Success. 

Since GeckoDriver/Firefox doesn't exhibit the symptom, it would appear the problem is with ChromeDriver. At least for my case, it simply doesn't play well with the X Server (XQuartz) and port forwarding over ssh on Mac. I did some poking around on their bug tracking site and found this:

https://bugs.chromium.org/p/chromedriver/issues/detail?id=2859&q=xquartz&colspec=ID%20Status%20Pri%20Owner%20Summary

So, I'm going to follow that path and see where it goes. 

**@aaltat**, thanks for your help and pointers.
